src/v/pandaproxy/client/client.cc

```diff
@@ -148,25 +153,33 @@ ss::future<kafka::fetch_response::partition> client::fetch_partition(
           return make_fetch_request(tp, offset, max_bytes, timeout);
       };
 
-    return ss::do_with(
-      std::move(build_request),
-      std::move(tp),
-      [this](auto& build_request, model::topic_partition& tp) {
-          return retry_with_mitigation(
-                   shard_local_cfg().retries(),
-                   shard_local_cfg().retry_base_backoff(),
-                   [this, &tp, &build_request]() {
-                       return _brokers.find(tp)
-                         .then([&tp, &build_request](shared_broker_t&& b) {
-                             return b->dispatch(build_request(tp));
+    return ss::with_gate(
+      _gate,
+      [this, build_request{std::move(build_request)}, tp{std::move(tp)}]() {
+          return ss::do_with(
+            std::move(build_request),
+            std::move(tp),
+            [this](auto& build_request, model::topic_partition& tp) {
+                vlog(ppclog.debug, "fetching: {}", tp);
+                return retry_with_mitigation(
```

Comment:
move to a searate method. this is very nested.

Reply:
question: how come these 2 configs aren't passed into the class and instead rely on this global state. 

the panda proxy should be able to be tested w/out manipulating the global configuration.

the main redpanda binary tries really hard to not do this kinda stuff and instead passes around PODs for configs to inner types.

Reply:
> move to a searate method. this is very nested.

Extracted `gated_retry_with_mitigation`.

> question: how come these 2 configs aren't passed into the class and instead rely on this global state.

Fair point. I will fix this in a separate PR. Issue: https://github.com/vectorizedio/redpanda/issues/179